source/common/stats/stat_data_allocator_impl.h

```diff
@@ -155,29 +164,36 @@ template <class StatData> class GaugeImpl : public Gauge, public MetricImpl {
   bool used() const override { return data_.flags_ & Flags::Used; }
 
   ImportMode importMode() const override {
-    if (data_.flags_ & Flags::ImportModeUninitialized) {
-      return ImportMode::Uninitialized;
+    if (data_.flags_ & Flags::NeverImport) {
+      return ImportMode::NeverImport;
     } else if (data_.flags_ & Flags::LogicAccumulate) {
       return ImportMode::Accumulate;
     }
-    return ImportMode::NeverImport;
+    return ImportMode::Uninitialized;
   }
 
   void mergeImportMode(ImportMode import_mode) override {
-    if (import_mode != ImportMode::Uninitialized &&
-        (data_.flags_ & Flags::ImportModeUninitialized)) {
-      data_.flags_ &= ~Flags::ImportModeUninitialized;
-      if (import_mode == ImportMode::Accumulate) {
-        data_.flags_ |= Flags::LogicAccumulate;
-      } else {
-        // A previous revision of Envoy must have transferred a gauge that it
-        // thought was Accumulate. But the new version thinks its NeverImport,
-        // so we clear the accumulated value.
-        data_.value_ = 0;
-        data_.flags_ &= ~Flags::Used;
-      }
-    } else {
-      ASSERT(importMode() == import_mode);
+    ImportMode current = importMode();
+    if (current == import_mode) {
+      return;
+    }
+
+    switch (import_mode) {
+    case ImportMode::Uninitialized:
```

Comment:
Can this ever happen given that this is the default on create? Should this be NOT_REACHED?

Reply:
Actually this line gets hit in the new testcase I wrote; this is the case where we set NeverImport on create, and then do a merge from a parent on the same name from a different scope.

Reply:
By different scope you mean the merge code, right? That's the only place where we initialize with uninitialized? Sorry this is kind of confusing, do you mind adding more comments?

Reply:
Yup, added:
```
      // mergeImportMode(ImportMode::Uninitialized) is called when merging an
      // existing stat with importMode() == Accumulate or NeverImport.
```